python_modules/dagster/dagster/core/storage/event_log/sql_event_log.py

```diff
@@ -652,152 +660,189 @@ def get_event_records(
-                .where(
-                    db.or_(
-                        AssetKeyTable.c.asset_key == asset_key.to_string(),
-                        AssetKeyTable.c.asset_key == asset_key.to_string(legacy=True),
-                    )
-                )
-                .where(
-                    db.or_(
-                        AssetKeyTable.c.wipe_timestamp == None,
-                        AssetKeyTable.c.last_materialization_timestamp
-                        > AssetKeyTable.c.wipe_timestamp,
-                    )
-                )
-                .limit(1)
-            )
-            with self.index_connection() as conn:
-                row = conn.execute(query).fetchone()
-                return bool(row)
+            # if we have migrated, we can limit using SQL
+            fetch_limit = limit
+        else:
+            # if we haven't migrated, overfetch in case the first N results are wiped
+            fetch_limit = max(limit, MIN_ASSET_ROWS) if limit else None
+        result = []
 
-        # has not migrated, need to pull asset_details to get wipe status
-        query = (
-            db.select([AssetKeyTable.c.asset_key, AssetKeyTable.c.asset_details])
-            .where(
-                db.or_(
-                    AssetKeyTable.c.asset_key == asset_key.to_string(),
-                    AssetKeyTable.c.asset_key == asset_key.to_string(legacy=True),
-                )
+        while should_query:
+            rows, has_more, current_cursor = self._fetch_raw_asset_rows(
+                asset_keys=asset_keys, prefix=prefix, limit=fetch_limit, cursor=current_cursor
             )
-            .limit(1)
-        )
+            result.extend(rows)
+            should_query = bool(has_more) and bool(limit) and len(result) < cast(int, limit)
 
-        with self.index_connection() as conn:
-            row = conn.execute(query).fetchone()
-            if not row:
-                return False
+        if self._can_mark_assets_as_migrated(rows, has_more):
+            self.enable_secondary_index(ASSET_KEY_INDEX_COLS)
 
-            asset_details: Optional[AssetDetails] = AssetDetails.from_db_string(row[1])
-            if not asset_details or not asset_details.last_wipe_timestamp:
-                return True
+        return result[:limit] if limit else result
 
-            materialization_row = conn.execute(
-                db.select([SqlEventLogStorageTable.c.timestamp])
-                .where(
-                    db.or_(
-                        AssetKeyTable.c.asset_key == asset_key.to_string(),
-                        AssetKeyTable.c.asset_key == asset_key.to_string(legacy=True),
-                    )
-                )
-                .order_by(SqlEventLogStorageTable.c.timestamp.desc())
-                .limit(1)
-            ).fetchone()
-            if not materialization_row:
-                return False
+    def _fetch_raw_asset_rows(self, asset_keys=None, prefix=None, limit=None, cursor=None):
```

Comment:
I think finding some way to break this method up would aid grokkability - it's a little difficult to tell which variables are only used within particular sections and which are used across sections.